lib/src/clojure_lsp/dep_graph.clj

```diff
@@ -0,0 +1,293 @@
+(ns clojure-lsp.dep-graph
+  (:require
+   [clojure-lsp.shared :as shared]
+   [clojure.set :as set]))
+
+;;;; Syntax
+
+;; () grouping
+;; | choice
+;; ? zero or one
+;; + one or more
+;; * zero or more
+;; '' literal
+;; "" string
+;; [] = list or vector
+;; {} = map {k1 v1 ...}
+;; #{} = set #{i1 ...}
+;; #<> = multiset #<i1 ...>
+;;       a mulitset is represented internally as a hashmap {i1 f1 ...} where f1 =
+;;       integer > 0, frequency of i1
+
+;; db = {':dep-graph' dep-graph
+;;       ':file-meta' file-meta}
+
+;; dep-graph = {(ns dep-graph-item)*}
+;; ns = symbol
+;; dep-graph-item =
+;; {(':dependencies' #<ns*>)?
+;;  (':dependents' #<ns*>)?
+;;  (':aliases' #<alias*>)?
+;;  (':files' #{filename*})?
+;;  (':internal?' boolean)?
+;;  (':from-internal?' boolean)?
+;;  (':from-langs' #{lang*})?}
+;; filename = string
+;; alias = (symbol | nil)
+;; lang = ':clj' | ':cljs' | ':edn'
```

Comment:
maybe bikeshedding, but it took me a little bit to understand this is describing the db from top-bottom, being top a higher level, maybe just invert to bottom-top?  describing lang first and then the things that use it until describing the db? maybe it's personal too

Reply:
I was sort of following the style of the Datomic [grammar](https://docs.datomic.com/cloud/query/query-data-reference.html#arg-grammar). But I've always found those docs a little hard to follow, partially because of your observation—the definitions are used before they're made. That's not how we do it in Clojure! :)

I'll take a stab at restructuring this. While I'm at it, do you want me to move it to `db/initial-db`? I kind of like it where it is, close to the code that needs to reference it most often, but I'd be OK with either location.

Reply:
I'm ok with either too, let's keep it here for now and see how it goes

Reply:
Grammar re-ordered; example added.

Reply:
looks great